src/res_core.ml

```diff
@@ -4049,46 +4075,51 @@ and parseTypeParameters p =
     let params =
       parseCommaDelimitedRegion ~grammar:Grammar.TypeParameters ~closing:Rparen ~f:parseTypeParameter p
     in
-    Parser.expect Rparen p;
+    Parser.expectUnsafe Rparen p;
     params
 
 and parseEs6ArrowType ~attrs p =
   let startPos = p.Parser.startPos in
   match p.Parser.token with
   | Tilde ->
-    Parser.next p;
+    Parser.expect Tilde p;
     let (name, loc) = parseLident p in
     let lblLocAttr = (Location.mkloc "ns.namedArgLoc" loc, Parsetree.PStr []) in
-    Parser.expect ~grammar:Grammar.TypeExpression Colon p;
+    Parser.expectUnsafe ~grammar:Grammar.TypeExpression Colon p;
     let typ =
       let typ = parseTypExpr ~alias:false ~es6Arrow:false p in
       {typ with ptyp_attributes = lblLocAttr::typ.ptyp_attributes}
     in
     let arg = match p.Parser.token with
     | Equal ->
-      Parser.next p;
-      Parser.expect Question p;
+      Parser.expect Equal p;
+      Parser.expectUnsafe Question p;
       Asttypes.Optional name
     | _ ->
       Asttypes.Labelled name
     in
-    Parser.expect EqualGreater p;
+    Parser.expectUnsafe EqualGreater p;
     let returnType = parseTypExpr ~alias:false p in
     let loc = mkLoc startPos p.prevEndPos in
     Ast_helper.Typ.arrow ~loc ~attrs arg typ returnType
   | _ ->
     let parameters = parseTypeParameters p in
-    Parser.expect EqualGreater p;
-    let returnType = parseTypExpr ~alias:false p in
-    let endPos = p.prevEndPos in
-    let typ = List.fold_right (fun (uncurried, attrs, argLbl, typ, startPos) t ->
-      let attrs = if uncurried then uncurryAttr::attrs else attrs in
-      Ast_helper.Typ.arrow ~loc:(mkLoc startPos endPos) ~attrs argLbl typ t
-    ) parameters returnType
-    in
-    {typ with
-      ptyp_attributes = List.concat [typ.ptyp_attributes; attrs];
-      ptyp_loc = mkLoc startPos p.prevEndPos}
+    if p.token <> Eof then (
```

Comment:
Does not look like it, it would need to go into `parseTypExpr`/`parseEs6ArrowType` loop, but that only works if there are fresh `=>` or `~` to look at.